onnxruntime/core/framework/execution_context.h

```diff
@@ -0,0 +1,190 @@
+  const std::vector<Stream*>& GetStreams() const;
+  size_t NumStreams() const;
+  Status CleanUp();
+
+ private:
+  std::unique_ptr<DeviceStreamCollectionImpl> impl_;
+};
+
+/*
+ * LIMITATION:
+ * CountDownBarrier is only for scenario that the v is set
+ * to the # of consumers and each consumer calls Dec() exactly once.
+ */
+class CountDownBarrier {
+ public:
+  CountDownBarrier() : v_{0} {};
+
+  void Set(int32_t v) {
+    ORT_ENFORCE(v >= 0);
+    v_.store(v, std::memory_order_relaxed);
+  }
+
+  bool Dec() {
+    return v_.fetch_sub(1, std::memory_order_relaxed) == 1;
+  }
+
+  int32_t Get() { return v_.load(std::memory_order_relaxed); }
+
+  void Inc() {
+    ++v_;
+  }
+
+ private:
+  std::atomic_int_fast32_t v_;
+};
+
+class SessionScope;
+typedef InlinedHashMap<std::string, OrtValue> OrtValueCache;
+typedef std::shared_ptr<OrtValueCache> OrtValueCachePtr;
+
+// execution context that support to execute a command on stream.
+// The notifications got instantiated when execution context is constructed.
+// TODO: if we merge the notifications to execution frame, we might don't need this.
+class ExecutionContext {
+ public:
+  ExecutionContext(const SessionState& sess_state,
+                   int32_t num_streams,
+                   gsl::span<const size_t> notification_owners,
+                   gsl::span<const int> feed_mlvalue_idxs,
+                   gsl::span<const OrtValue> feeds, gsl::span<const int> fetch_mlvalue_idxs,
+                   std::vector<OrtValue>& fetches,
+                   const InlinedHashMap<size_t, IExecutor::CustomAllocator>& fetch_allocators,
+                   size_t num_barriers,
+                   const logging::Logger& sess_logger,
+                   const DeviceStreamCollection& device_stream_map,
+                   bool single_thread_mode);
+
+  const SessionState& GetSessionState() const;
+
+  const logging::Logger& GetLogger() const;
+
+  ExecutionFrame* GetExecutionFrame();
+
+  synchronize::Notification* GetNotification(size_t idx);
+
+  const bool* TerminateFlag() const;
+
+  void SetLogger(const logging::Logger& current_logger) {
+    logger = &current_logger;
+  }
+
+  void SetTerminateFlag(const bool* terminate_flag) {
+    terminate_flag_ = terminate_flag;
+  }
+
+  const Status& TaskStatus() const;
+
+  bool DecCountDownBarrier(size_t barrier_id);
+
+  bool SingleThreadMode() const { return single_thread_mode_; }
+
+  Stream* GetDeviceStream(size_t idx);
+
+  void CompleteTask();
+
+  void AddTask();
+
+  void WaitAll();
+
+  void SetStatus(Status& status);
+
+  void RecycleNodeInputs(onnxruntime::NodeIndex node_index);
+
+#ifdef ENABLE_TRAINING
+  void SetOrtValueCache(OrtValueCachePtr cache) {
+    cache_ = std::move(cache);
+  }
+
+  OrtValueCachePtr& GetOrtValueCache() {
+    return cache_;
+  }
+#endif
+
+  ~ExecutionContext();
+
+  SessionScope* GetSessionScope() {
+    return session_scope_;
+  }
+
+  void SetSessionScope(SessionScope* session_scope) {
```

Comment:
If the context should always have a session scope, but that might get changed later, it would be cleaner to provide an initial session scope in the ctor and allow updating it later.

If session_scope should not be a nullptr please use a reference for the param. 

If we have an initial session scope in the ctor and the method to update it that takes a reference that would enable an invariant where the session scope should always be valid and GetSessionScope can return a reference.

Reply:
@RandySheriffH

Reply:
fixed by pass in SessionScope as a standalone argument during execution.